Speed up grid_data endpoint by 10x

ashb opened the pull request:
These changes make the endpoint go from almost 20s down to 1.5s and the
changes are two fold:

1. Keep datetimes as objects for as long as possible

   Previously we were converting start/end dates for a task group to a
   string, and then in the parent parsing it back to a datetime to find
   the min and max of all the child nodes.

   The fix for that was to leave it as a datetime (or a
   pendulum.DateTime technically) and use the existing
   `AirflowJsonEncoder` class to "correctly" encode these objects on
   output.

2. Reduce the number of DB queries from 1 per task to 1.

   The removed `get_task_summaries` function was called for each task,
   and was making a query to the database to find info for the given
   DagRuns.

   The helper function now makes just a single DB query for all
   tasks/runs and constructs a dict to efficiently look up the ti by
   run_id.

ashb committed:
Speed up grid_data endpoint by 10x

These changes make the endpoint go from almost 20s down to 1.5s and the
changes are two fold:

1. Keep datetimes as objects for as long as possible

   Previously we were converting start/end dates for a task group to a
   string, and then in the parent parsing it back to a datetime to find
   the min and max of all the child nodes.

   The fix for that was to leave it as a datetime (or a
   pendulum.DateTime technically) and use the existing
   `AirflowJsonEncoder` class to "correctly" encode these objects on
   output.

2. Reduce the number of DB queries from 1 per task to 1.

   The removed `get_task_summaries` function was called for each task,
   and was making a query to the database to find info for the given
   DagRuns.

   The helper function now makes just a single DB query for all
   tasks/runs and constructs a dict to efficiently look up the ti by
   run_id.

potiuk reviewed (approved):
:heart: it

norm reviewed (approved):
Love it.

dstandish commented on airflow/utils/json.py:
```diff
@@ -45,7 +45,7 @@ def __init__(self, *args, **kwargs):
         """Convert dates and numpy objects in a json serializable format."""
         if isinstance(obj, datetime):
-            return obj.strftime('%Y-%m-%dT%H:%M:%SZ')
+            return obj.isoformat()
```
this seems to produce different output:
```
pendulum.now().timestamp()
1654540994.520137
from datetime import datetime
now = datetime.utcnow()
now.strftime('%Y-%m-%dT%H:%M:%SZ')
'2022-06-07T16:05:10Z'
now.isoformat()
'2022-06-07T16:05:10.271025'
```
do you think it matters?

ashb replied:
`Z` vs `+00:00` shouldn't matter to any compliant parser. So this should be okay

ashb replied:
I've just changed this (to fix tests) so that it uses the existing `is_naive` and `convert_to_utc` functions so it will always include a TZ.

dstandish replied:
But the original code also stripped out sub-second precision and isoformat includes micros, it appears. Don't know whether that matters here.

ashb replied:
Oh right -- it shouldn't.

bbovenzi commented on airflow/www/views.py:
```diff
@@ -250,62 +251,102 @@ def _safe_parse_datetime(v):
+                    'state': task_instance.state,
+                    'start_date': task_instance.start_date,
+                    'end_date': task_instance.end_date,
+                    'try_number': try_count,
```
This no longer returns the `mapped_states` to give a status breakdown of mapped tasks. Do we still want that? If not, I need to change one bit of react code.

ashb replied:
Oh yes. Removing that was not intentional!

ashb replied:
Where did we include that before? I'm not seeing where I delete it in this diff? (My brain is foggy though)

ashb replied:
Oh

```diff
-        if task_instance.map_index > -1:
-            return get_mapped_summary(
-                task_instance, task_instances=get_mapped_instances(task_instance, session)
-            )
```

bbovenzi replied:
Everything else seems to be fine we just need something like `mapped_states`, best format would be like so (I'll make any UI changes needed):

```
{
  success: 3.
  running: 2,
  queued: 1,
}
```

ashb replied:
Do you even need the counts, or can we make the decision of what state to show on the server side?

bbovenzi replied:
We can make the decision on the server. We were giving an overall status and a breakdown of the status of all the children that form that overall status. But we can just do an overall status if you'd like

ashb replied:
Oh showing the breakdown in the tooltip? If so I'll leave the counts back

ashb replied:
Right, made this change, and updated the React to match in the two places I found. Did I miss anywhere @bbovenzi

ashb committed:
Add support for mapped tasks in the grid data

ashb committed:
Don't fail when not all tasks have a finish date.

Note that this possibly has incorrect behaviour, in that the end_date of
a TaskGroup is set to the max of all the children's end dates, even if
some are still running. (This is the existing behaviour and is not
changed or altered by this change - limiting it to just performance
fixes)

bbovenzi committed:
fix eslint errors
